### PR TITLE
REST API for asynchronous payload callbacks

### PR DESCRIPTION
**Pull request by asoto-r7**
## WIP: Ready for testing and feedback.

This addition to the data services and web services adds support for asynchronous payloads (aka "pingback payloads") to be tracked, and is meant to coincide with a [matching pull request to the metasploit data model](https://github.com/rapid7/metasploit_data_models/pull/179).

## Verification

0. If https://github.com/rapid7/metasploit_data_models/pull/179 has not yet been merged, then you must first complete the `Landing` section contained within that PR.  Use the `Verification` instructions within that same PR to confirm that all functionality works as expected.

1. Run `./msfdb reinit`, walk through the steps, and copy the security token to your clipboard:

```
$ ./msfdb reinit
[...]
MSF web service username: postgres
MSF web service password: [...]
MSF web service user API token: 25acf47a21b2a7b1f18f7a45e13cdd32ae6e1200bcd9be75dc294e31423e05becacfe7b6431db95b
```

2. Start the database and web services: `./msfdb start`.  Feel free to append `-a 0.0.0.0` to make the web server externally accessible, eg. if you'll be using a browser outside the VM.

3. Load Swagger by browsing to `https://127.0.0.1:5443/api/v1/api-docs`.  Note, I had to use Chrome on Windows, since Firefox in Ubuntu wasn't loading the page.  Also, note that the page loads external resources, so you may need to configure browser security plugins and confirm your test environment is Internet-connected.

4. In the top-right corner, click the `Authenticate` button

image

5.  Paste in the token from Step 1 and click "Authorize".

image

6. Scroll down to the first header, `async_callback` and confirm that three endpoints are shown: GET, POST, and GET:

image

7. Expand the first endpoint, `GET ​/api​/v1​/async-callbacks`, click the `Try It Out` button, enter `default` for the Workspace, and click the Execute button:

image

8. Confirm that the first endpoint returned an empty `data` array (since we cleared the database in Step 1):

image

8. Expand the second endpoint, `POST ​/api​/v1​/async-callbacks`, click the `Try It Out` button, confirm that `default` is used for the Workspace, and click the Execute button:

image

9. Repeat step 7, and confirm that your callback information is returned in the data array.

10. Expand the third endpoint, `GET ​/api​/v1​/async-callbacks/{uuid}`, click the `Try It Out` button, enter the UUID from above and `default` as the Workspace, then click the Execute button:

image

11. Test the second POST endpoint again with a new URL.

12. Test both GET endpoints and confirm they show all callbacks that meet your search criteria.

## Random troubleshooting steps

If Swagger returns a TypeError, confirm that the `msfdb` web server started.  You should see it listening in your `netstat` on port 5443.  If it's running, then refresh the Swagger page, as your SSL cert may have been regenerated and your requests are failing in the background.

If you get back an error within Swagger, like an Internal Server Error, check your `~.msf4/msf-ws.log`.

If `msfdb` doesn't start, use the `-d` flag to enable debugging, then run the commands manually.

### CODE REVIEW

**Comment by busterb** [comment]
This looks good to me. @mkienow-r7 / @ebleiweiss-r7 / @jbarnett-r7 any comments on the approach taken? Does this look reasonable from a data model / API implementation PoV?

**Comment by mkienow-r7** [comment]
Would you clarify the following step. Is this asking one to create another POST where "new URL" is a new `listener_uri`?
> 11. Test the second POST endpoint again with a new URL.

**Comment by mkienow-r7** [comment]
Would you elaborate on the need for the new table and DBManager? At a quick glance this seem very similar to the changes introduced by @jbarnett-r7 in rapid7/metasploit_data_models#172.

File: db/schema.rb [hunk 1]
```diff
@@ -22,6 +22,18 @@
...
+  create_table "async_callbacks", force: :cascade do |t|
+    t.string   "uuid",           null: false
+    t.integer  "timestamp",      null: false
+    t.string   "listener_uri"
+    t.string   "target_host"
+    t.string   "target_port"
+    t.integer  "workspace_id"
+    t.datetime "created_at",     null: false
+    t.datetime "updated_at",     null: false
+    t.uuid     "{:null=>false}"
```
Comment (jbarnett-r7) [inline db/schema.rb]: Isn't this line redundant with line 26?

**Comment by mkienow-r7** [comment]
The Swagger documentation for `POST ​/api​/v1​/async-callbacks` shows use of a `workspace_id` field as the name of the workspace. This should be `workspace`, as seen in the host, loot, note, etc. models.

File: lib/msf/core/db_manager/async_callback.rb [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        #return Array.wrap( Mdm::AsyncCallback.where ( uuid: opts[:uuid] ))
+        return Mdm::AsyncCallback.where( 'uuid' => opts[:uuid] )
+      end
+
+      wspace = Msf::Util::DBManager.process_opts_workspace(opts, framework)
+      return wspace.async_callbacks.where(opts)
+    end
+  end
+
+  def update_async_callback(opts)
```
Comment (mkienow-r7) [inline lib/msf/core/db_manager/async_callback.rb]: Why does the `DBManager` expose an update method that isn't usable from these additions?

**Comment by mkienow-r7** [comment]
Have these additions been tested from code within framework, that is, making calls to the new `DBManager` methods (`framework.db.create_async_callback`, `framework.db. async_callbacks`, `framework.db.update_async_callback`) while using both a local and remote data service connection? I see the DataProxy, remote DataService and stub modules, however, there is no associated entry for these additions in the `DataProxyAutoLoader` module (lib/metasploit/framework/data_service/proxy/data_proxy_auto_loader.rb) or `DataServiceAutoLoader` (lib/metasploit/framework/data_service/remote/http/data_service_auto_loader.rb). These missing changes will cause issues similar to what was resolved in #11532.

**Comment by mkienow-r7** [comment]
Minor: Would it be helpful if the naming had a more clear indication that these new components are related to payloads? Perhaps "payload callbacks" instead of "async callbacks"?

**Comment by asoto-r7** [comment]
@mkienow-r7 @jbarnett-r7 : Thanks for the feedback.  We thought we wanted to have an external listener use this REST API to report back, but we've found that integrating directly with framework is easier and more flexible.  So, I'm going to close this PR, but this feedback will be helpful as we extend into having user APIs that will need a similar mechanism.

So, thanks everyone!

**Pull request closed without merging**